### ISSUE TITLE
SPFx WebPart does randomly not show on SharePoint page and seems not loaded

### ISSUE THREAD

**Issue by ThomasContini** [SharePoint/sp-dev-docs#9062]
### Target SharePoint environment

SharePoint Online

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Developer environment

Windows

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [X] 💥 Microsoft Edge
- [ ] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [ ] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- Microsoft Edge v. 114.0.1823.58
- SPFx 1.13.1
- Node.js 14.19.0

### Describe the bug / error

We developed a custom webpart in order to make a kind of advanced toolbar. This webpart has been deployed to all SharePoint sites from a specific hub (through site collection app catalogs), and installed on every sites' home pages.

The component is placed at the very top of a vertical section.

Most of the time, the component is perfectly working, and shows correctly. However, while navigating between pages through hub navigation, and completely randomly, it get not loaded correctly.

### Logs

No issues in console from the component itself. Only SP issues that we already saw on other tenants or pages... I still let you the complete log:

![console1](https://github.com/SharePoint/sp-dev-docs/assets/80830836/722605d7-7d49-4159-aea2-ef0bbf2cf0ef)

![console2](https://github.com/SharePoint/sp-dev-docs/assets/80830836/ad67451a-6add-4cce-9c5c-51a1617df68c)

### DOM

It seems like the component is not even loaded by SharePoint Framework! DOM is empty at component location:

![DOM-when-error](https://github.com/SharePoint/sp-dev-docs/assets/80830836/e5a40fa4-7383-4ae8-b919-2cdbf5ee641c)

### Site sources

I decided to look at the loaded sources thanks to the developer console in Edge. The client side assets are not even present in sources when this bug is appearing:

#### When component **does not** correctly:

![Assets-when-error](https://github.com/SharePoint/sp-dev-docs/assets/80830836/4052f24f-981a-41b5-99ff-3a6c8c2bcc5f)

#### When component shows correctly:

![Assets-when-good](https://github.com/SharePoint/sp-dev-docs/assets/80830836/8d5ce6d3-3a0e-4e0c-a956-4c20e52ea92b)

We have other components deployed on the same pages, but they show correctly. However, the same bug has already been seen on other custom SPFx components, still randomly, but less time than this one I would say.

### Recurrence

This bug happens for all end users and developers working on this specific tenant. 
It still seems more redundant as end user, without admin rights.

### My own thinking

- Could it be related to the fact that this tenant seems really old (it get created a long time ago...)?
- Is it possible that my code generates an error that crashes the entire bundle even before it get loaded by SharePoint?
- Am I missing something? I can't explain why this one does not show up every time, and it's difficult to accept that the bug is probably generated by my code as it seems not even loaded...

Hope I provided enough information :)

### Steps to reproduce

This one is quite tricky to reproduce. We are currently not able to reproduce it on other tenants with the **same** component in the same configuration. It seems only happening on a specific tenant.

Here is how we reproduce it on this specific tenant:

1. Create multiple site collections, create site collection app catalogs on it, then deploy a custom component
2. Add all these new sites to a new hub, then update hub navigation to be able to navigate between home pages 
3. Install this component on every sites' home pages, configure it in a really similar way 
4. Navigate between home pages through the Hub navigation, the bug could appear after multiple retries. We currently need about 10-20 attempts to reproduce it...

Sorry to not be able to provide more infos about how you can reproduce this issue...

### Expected behavior

The component should always show itself correctly.

**Comment by AJIXuMuK** [comment]
@ThomasContini if you look closely to the console logs you will see that your custom service is failing due to CORS configuration.
I assume, you're making a request in `onInit` -> request is failing -> web part is not rendering.

**Comment by ThomasContini** [comment]
Hi @AJIXuMuK,

Thanks for your feedback!

Unfortunately this CORS issue is not issued from my component code. Moreover, this issue appears everytime we use [PnP controls](https://pnp.github.io/sp-dev-fx-controls-react/) because of their telemetry but should not stop code execution, as we see this in every single component we developed using [PnP controls](https://pnp.github.io/sp-dev-fx-controls-react/), and they are still loading... 

In order to prove that, I just made a new attempt, bug still reproduces, this time no error about CORS policy and still the component is not showing:

![rep1-console1](https://github.com/SharePoint/sp-dev-docs/assets/80830836/4d673aa0-76bf-4e97-9745-6b562ddb230d)
![rep1-console2](https://github.com/SharePoint/sp-dev-docs/assets/80830836/e7078888-b587-48c3-a161-3b6c7bc0f592)

I also made a new attempt by opting out PnP telemetry, issue is still here.

IMHO, If my `onInit` method was crashing, I would get a different error such as the famous `Something went wrong`, because it would not reach my renderer.

**Comment by AJIXuMuK** [comment]
@ThomasContini - sorry for the delay.

> IMHO, If my onInit method was crashing, I would get a different error such as the famous Something went wrong, because it would not reach my renderer.

It depends. If you are throwing the error from `onInit` or rejecting the promise then yes. But if you just "swallow" the exception - it will not be displayed.

Could you please check if the same issue occurs on some dummy/simple web part on the same page/tenant?
Could you share your solution or test tenant with me?

You can use aterentiev@microsoft.com to share it.

Thanks!

**Comment by ThomasContini** [comment]
Hi @AJIXuMuK, 

I just invite you on the tenant, and sent you URL to connect by email. 
I also shared you the source code through a private Github repo.

Hope you have everything that could be useful to understand the issue.

Thanks!

**Comment by andrewpeterson99** [comment]
Could I ask what the resolution to this was? I'm running into a similar issue.

**Comment by heesungjang** [comment]
Hi @ThomasContini  / @AJIXuMuK ,

When you have some time, could you please share a bit of background on what the underlying issue was and how it was resolved?

We are experiencing the same issue. One custom web part is disappearing from the SharePoint page during site navigation. As @ThomasContini  mentioned earlier in this ticket, it seems like the component is not even loading—the DOM is empty at the component’s location. This issue is particularly challenging because there is no distinct error message we can trace.

Appreciate it in advance.

**Comment by heesungjang** [comment]
@andrewpeterson99 , 

Hi Andrew,

I’m not sure if this will help resolve the issue you’re experiencing. In our case, we saw the following error message in the console when the WebPart intermittently disappeared:

[Failed to execute 'removeChild' on 'Node']

For us, the issue was resolved by explicitly unmounting the component before it calls render function.

Below is a snippet from our code:

<img width="409" alt="Image" src="https://github.com/user-attachments/assets/5900b4c8-6a7c-4b72-a198-698f408bc10c" />

**Comment by dhiabedoui** [comment]
Hi @heesungjang I’m investigating a related issue where my SPFx web part’s content disappears unexpectedly after SPA navigation between SharePoint sites. I’ve narrowed it down to the usage of ClientsidePageFromFile from PnPjs (version 3.23.0). When this method is used to fetch the current page (getFileByServerRelativePath(pageContext.site.serverRequestPath)), the content of the web part renders briefly and then vanishes and the container remains in the DOM but is emptied.

Are you using ClientsidePageFromFile in your web part or extension? Or any related methods like getFileByServerRelativePath() ?

**Comment by stopacio** [comment]
Running into something similar as well, but my webpart only retrieves Site Pages and some List views.  It isn't consistent enough to figure out what's causing it, but I do notice that when it happens, it also doesn't show the PnP Modern Search Web Part I have just below it but it doesn't kill another custom web part that uses an older version of SPFx (1.14) and related PnP dependencies ( @pnp/sp@4.0.1, @pnp/spfx-controls-react@3.17.0,@pnp/spfx-property-controls@3.16.0), so I wonder if there's a connection?

**Comment by EkiThomasContini** [comment]
Hi,

First of all, sorry for this late (very (VERY) late...) answer.

Unfortunately, we never really resolved this particular issue. We finally managed by refactor the whole WebPart and reduce the amount of requests we made to the SharePoint API. Former code implementation was requesting about 6-7 SP lists through REST API in the WP `onInit` method. New code is only making 2-3 requests, and results are cached inside local storage, and at dev time, we never got this issue again.

At this time, we don't have any explanation about this previous issue. We previsously thought it could be related to throttling: that's why we tried a huge refactor (that is now working). 

IMHO, throttling cannot explain JS bundle not even being visible though the developer console...

**Comment by EkiThomasContini** [comment]
We are now running into similar issues for few weeks:
- Some of our WebParts do not load on page ;
- They show up after refreshing the page (F5 or ctrl+F5) multiple times. 

The issue is slightly different, we can see JS bundles in developer mode, but SPFx `div` wrapper is totally empty.
Source code seems to be called as the dev console prints some warning from "**slider-web-part**" (our concerned WP).

**As mentionned before, source files now appear:**
<img width="627" height="250" alt="Image" src="https://github.com/user-attachments/assets/4d045535-da06-4451-9a9d-e44054091534" />

**Div host is empty**
<img width="1129" height="173" alt="Image" src="https://github.com/user-attachments/assets/267d863d-348a-4044-b89b-e4d91af9ee6b" />

**Code seems to be executed as suggests logs**
<img width="1882" height="596" alt="Image" src="https://github.com/user-attachments/assets/d0c4761c-e262-4fb4-af6b-0d86e7200006" />

In this specific case, `onInit` method:
-  is deterministic ;
- contains REST calls but encapsulated in try/catch logic ;

However the WebPart is randomly not rendering as user screen... A simple refresh (F5) fix the issue in most (_but not all_) cases.

**Comment by AmarNathKovuri** [comment]
@andrewpeterson99  got any fix

**Comment by bcameron1231** [comment]
@AJIXuMuK I am seeing the same issue. Happening during navigation events (not refreshes). 

Render is called a few times. The first few times the this.domElement has an innerHTML. On the render when it fails, this.domElement has no innerHTML as if the component was unmounted.

I've been trying to track down the issue to no avail.

**Comment by bcameron1231** [comment]
I investigated a bit more. 

From my side, the web part initializes correctly on a normal hard refresh, but when navigating to the page through SharePoint’s intercepted navigation, the web part can begin rendering and then its root DOM is later cleared by the SharePoint/SPFx loading pipeline for the same instance (`_createLoadingIndicatorElement`) clears the children of the dom). The container remains on the page, but the web part content disappears and the root ends up empty.

I traced this to the SharePoint Pages reinitialize/navigation flow. During the delayed load scenario, SPFx calls the loading indicator/status renderer again for the same web part instance, and that path clears the web part root. This seems to happen while the web part is still in a longer async onInit method. It's pretty easy to reproduce by sticking a breakpoint in the code to delay it.

As a workaround, I added a local initialization guard to prevent the web part from rendering before the onInit fully completes. That stabilizes the behavior, but from my perspective the underlying issue appears to be in the SharePoint/SPFx page navigation and render lifecycle rather than in the web part logic itself.
